common/message_passing/client.h

```diff
@@ -0,0 +1,154 @@
+   *    immediately, before the message is sent. Errors will not be reported.
+   * @param message The message to send.
+   * @return True if it succeeded in dispatching the send request, false
+   *    otherwise.
+   */
+  bool SendAsync(const google::protobuf::Message& message);
+
+  /**
+   * @brief Receives a message.
+   * @tparam MessageType The type of message we expect.
+   * @param message[out] The message that we received.
+   * @return True if a message was successfully received and parsed, false
+   *    otherwise.
+   */
+  template <class MessageType>
+  bool Receive(MessageType* message) {
+    // Make sure we're connected.
+    EnsureConnected();
+
+    // Create the parser for the response.
+    wire_protocol::MessageParser<MessageType> parser;
+
+    // Receive the response.
+    while (!parser.HasCompleteMessage()) {
+      const auto kResponse = receive_queue_->Pop();
+      if (kResponse.status <= 0) {
+        // The receive failed or the endpoint disconnected.
+        return false;
+      }
+
+      parser.AddNewData(kResponse.message);
+    }
+
+    return parser.GetMessage(message);
+  }
+
+  /**
+   * @brief Sends a request, then waits for a response from the receiver.
+   * @tparam ResponseType The type of response message we expect.
+   * @param request The request to be sent.
+   * @param response[out] The response that we received.
+   * @return True if the request was successfully sent and a response was
+   *    successfully received, false otherwise.
+   */
+  template <class ResponseType>
+  bool SendRequest(const google::protobuf::Message& request,
+                   ResponseType* response) {
+    // First, send the request.
+    if (Send(request) < 0) {
+      return false;
+    }
+
+    // Now, receive the response.
+    return Receive(response);
+  }
+
+ private:
+  /**
+   * @brief Dispatches a new message send operation.
+   * @param message The message to send.
+   * @param async Whether to send the message asynchronously.
+   * @param id[out] Filled with the chosen message ID, if not null.
+   * @return True if the dispatch succeeded, false otherwise.
+   */
+  bool DispatchSend(const google::protobuf::Message& message, bool async,
+                    MessageId* id);
+
+  /**
+   * @brief Ensures that we are connected to the server.
+   * @return True if it is connected, false if it failed to connect.
+   */
+  bool EnsureConnected();
+
+  /// The thread pool to use for running tasks.
+  std::shared_ptr<thread_pool::ThreadPool> thread_pool_;
+  /// The endpoint we are sending messages to.
+  Endpoint endpoint_;
+  /**
+   * @brief Queue containing messages for the sender task to send.
+   */
+  std::shared_ptr<queue::Queue<SenderTask::SendQueueMessage>> send_queue_;
+  /// Queue containing received messages.
+  std::shared_ptr<queue::Queue<ReceiverTask::ReceiveQueueMessage>>
+      receive_queue_;
+  /// The task responsible for sending messages.
+  std::shared_ptr<SenderTask> sender_task_ = nullptr;
```

Comment:
Is there a reason you initialized them to `nullptr` rather than leaving them uninitialized?

Reply:
Sort of a defensive programming thing. If you initialize them to `nullptr` and try to do something with them, it's *guaranteed* to segfault as opposed to being potentially-undefined behavior. Obviously, this situation should never happen, and I'd rather catch it quickly if it ever does.